### PR TITLE
Rough draft of a p2p noise protocol implementation

### PR DESCRIPTION
**Pull request by vtnerd**
My CCS is ending, so I wanted to post a draft of the noise protocol because that has been one of the primary efforts. I also wanted some final feedback if the Noise protocol (vs SSL) is worth doing, because I really have to hack the TCP server a bit for this to work.

The code is almost finalized - some stuff is needed on the detection end. I mainly wanted to highlight the painful hacks to squeeze this in.

### CODE REVIEW

**Comment by jeffro256** [comment]
Does the noise code currently save any keys persistently like the SSL files?

**Comment by jeffro256** [comment]
There may be a benefit to integrating an existing library (e.g. https://github.com/rweather/noise-c) versus writing the crypto code from scratch in that it will be less likely that our usage of noise is fingerprinted for some reason due to an oversight. Don't get me wrong, it would still be a lot of work hacking it into the TCP server, but are there other reasons you haven't used an existing library?

File: contrib/epee/include/net/connection_basic.hpp [hunk 1]
```diff
@@ -58,7 +58,6 @@ namespace epee
 {
 namespace net_utils
 {
-
```
Comment (jeffro256) [inline contrib/epee/include/net/connection_basic.hpp:61]: unnecessary whitespace changes

File: contrib/epee/include/net/levin_base.h [hunk 2]
```diff
@@ -71,7 +71,6 @@ namespace levin
   };
 #pragma pack(pop)
 
-
```
Comment (jeffro256) [inline contrib/epee/include/net/levin_base.h:74]: another one

File: contrib/epee/include/net/net_ssl.h [hunk 3]
```diff
@@ -45,11 +45,12 @@ namespace epee
...
 namespace net_utils
 {
+  // ssl_support_t
 	enum class ssl_support_t: uint8_t {
 		e_ssl_support_disabled,
 		e_ssl_support_enabled,
-		e_ssl_support_autodetect,
-	};
+		e_ssl_support_autodetect
+    };
```
Comment (jeffro256) [inline contrib/epee/include/net/net_ssl.h:53]: more whitespace changes plus comment that prob isn't needed

**Review by jeffro256 — COMMENTED** [review]
You could knock 3 of these files off the file modifications list by undoing whitespace changes

**Comment by vtnerd** [comment]
> Does the noise code currently save any keys persistently like the SSL files?

No, all of the keys are generated per-stream and never re-used. Initially my writeup had static keys, but I was concerned about fingerprinting via those keys. In a future version upgrade, static keys can be added opt-in; I think automagically doing it for monero p2p is the wrong approach.

> There may be a benefit to integrating an existing library (e.g. https://github.com/rweather/noise-c) versus writing the crypto code from scratch in that it will be less likely that our usage of noise is fingerprinted for some reason due to an oversight. Don't get me wrong, it would still be a lot of work hacking it into the TCP server, but are there other reasons you haven't used an existing library?

I wanted more control over the protocol which may have been overkill - the length fields are also encrypted+tagged in my proposal. I might be able to get this working with noise-c but it's convoluted even with raw libsodium functions - I have to decrypt the header without auth first before doing the proper decrypt+authenticate call. This is probably overkill because you can get a rough idea of the message size based on timing analysis (one of the reasons why I plan to add a padding feature to all of this).

The noise protocol is somewhat simple to implement, and this was weighed against adding yet-another-dependency that implements all of the Noise variants (Monero should only need two of the variants). And I'm still only about 75% certain that using Noise is worth the effort over SSL/TLS.

**Comment by vtnerd** [comment]
Closing, since the P2P SSL branch was done instead.
